crates/pallet-transporter/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+        /// Currency used by this pallet.
+        type Currency: Currency<Self::AccountId>;
+    }
+
+    /// Pallet transporter to move funds between domains.
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    /// Stores the next outgoing transfer nonce.
+    #[pallet::storage]
+    #[pallet::getter(fn next_outgoing_transfer_nonce)]
+    pub(super) type NextOutgoingTransferNonce<T: Config> =
+        StorageMap<_, Identity, T::DomainId, Nonce, ValueQuery>;
+
+    /// All the outgoing transfers on this execution environment.
+    #[pallet::storage]
+    #[pallet::getter(fn outgoing_transfers)]
+    pub(super) type OutgoingTransfers<T: Config> = StorageDoubleMap<
+        _,
+        Identity,
+        T::DomainId,
+        Identity,
+        Nonce,
+        Transfer<T::DomainId, T::AccountId, BalanceOf<T>>,
+        OptionQuery,
+    >;
+
+    /// Stores the next incoming transfer nonce.
+    #[pallet::storage]
+    #[pallet::getter(fn next_incoming_transfer_nonce)]
+    pub(super) type NextIncomingTransferNonce<T: Config> =
+        StorageMap<_, Identity, T::DomainId, Nonce, ValueQuery>;
+
+    /// All the incoming transfers on this execution environment.
+    #[pallet::storage]
+    #[pallet::getter(fn incoming_transfers)]
+    pub(super) type IncomingTransfers<T: Config> = StorageDoubleMap<
+        _,
+        Identity,
+        T::DomainId,
+        Identity,
+        Nonce,
+        Transfer<T::DomainId, T::AccountId, BalanceOf<T>>,
+        OptionQuery,
+    >;
+
+    /// Events emitted by pallet-transporter.
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Emits when there is a new outgoing transfer.
+        OutgoingTransfer {
+            /// Destination domain the transfer is bound to.
+            domain_id: T::DomainId,
+            /// Nonce of the transfer.
+            nonce: Nonce,
+        },
+    }
+
+    /// Errors emitted by pallet-transporter.
+    #[pallet::error]
+    pub enum Error<T> {
+        LowBalance,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Initiates transfer of funds from account on src_domain to account on dst_domain.
+        /// Funds are burned on src_domain first and are minted on dst_domain using Messenger.
+        #[pallet::weight((10_000, Pays::No))]
+        pub fn transfer(
+            origin: OriginFor<T>,
+            dst_location: Location<T::DomainId, T::AccountId>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            T::Currency::withdraw(
+                &sender,
+                amount,
+                WithdrawReasons::TRANSFER,
+                ExistenceRequirement::AllowDeath,
+            )
+            .map_err(|_| Error::<T>::LowBalance)?;
+
+            let dst_domain_id = dst_location.domain_id;
```

Comment:
Now `dst_domain_id` can be same with `src_domain_id`, should we also check `dst_domain_id` is different from SelfDomainId and only support the transfers between the different domains otherwise it's just a regular transfer which can be done in pallet-balances? If true, I'm wondering if it's worthy calling `Transfer` as `CrossDomainTransfer` to make it explicit and emphasize it's a cross domain operation.